python/lsst/daf/persistence/repository.py

```diff
@@ -29,231 +29,107 @@
+            self.mode)
+
+    @property
+    def cfgRoot(self):
+        if self._cfgRoot is not None:
+            return self._cfgRoot
+        else:
+            return self.root
 
-    # todo these load & write methods are coupled to posix storage. need to invent butler mechanism for
-    # multiple dispatch and implement it.
     @staticmethod
-    def butlerRead(butlerLocation):
-        if butlerLocation.getStorageName() is not "YamlStorage":
-            raise NotImplementedError("RepositoryCfg only supports YamlStorage")
-        ret = []
-        for location in butlerLocation.getLocations():
-            logLoc = LogicalLocation(location, butlerLocation.getAdditionalData())
-            with open(logLoc.locString()) as f:
-                cfg = yaml.load(f)
-            cfg['accessCfg.storageCfg.root'] = os.path.dirname(location)
-            ret.append(cfg)
-        return ret
+    def inputRepo(storage, tags=None):
+        return RepositoryArgs(storage, tags)
 
     @staticmethod
-    def butlerWrite(obj, butlerLocation):
-        if butlerLocation.getStorageName() is not "YamlStorage":
-            raise NotImplementedError("RepositoryCfg only supports YamlStorage")
-        ret = []
-        for location in butlerLocation.getLocations():
-            logLoc = LogicalLocation(location, butlerLocation.getAdditionalData())
-            if not os.path.exists(os.path.dirname(logLoc.locString())):
-                os.makedirs(os.path.dirname(logLoc.locString()))
-            with open(logLoc.locString(), 'w') as f:
-                yaml.dump(obj, f)
+    def outputRepo(storage, mapper=None, mapperArgs=None, tags=None, mode=None):
+        return RepositoryArgs(storage, mapper, mapperArgs, tags, mode)
 
-class Repository(object):
-    """
-    Default Multiple Parent & Output Behaviors.
-    * put/write operates only in non-parent repositories.
-    * get/read operates only in parent repositories.
-    Multiple output support:
-    * a Repository can have peer repositories.
-    * all outputs (writes) go to all non-parent repositories. (e.g. mapping with write==False will return
-      mappings for from all peer Repositories.
-    Multiple parent support:
-    * parents are processed in priority order, as defined by the order of the tuple passed in as
-      cfg['parentCfgs']
-    * parent search is depth first (1st parent, parents of 1st parent, grandparents of 1st parent, 2nd parent,
-      etc)
-    * if parentJoin is 'left' returns after first result is found.
-    * if parentJoin is 'outer' will return all results found.
-
-    Recursion is implemented in a few functions which may be overridden to alter recursive behavior:
-    def doPeersAndParents()
-    def doPeers()
-    def doParents()
+    def tag(self, tag):
+        """add a tag to the repository cfg"""
+        if isinstance(tag, basestring):
+            self.tags.add(tag)
```

Comment:
Should this be `if isinstance(tag, set)`?  I don't think `add` is a method of `basestring`.